File: Remove Nth Node From End of List/Remove Nth Node From End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode ptr1 = head;
        ListNode ptr2 = head;
        
        if(n == 0){
            return head;
        }
        
        for(int i = 0; i < n; ++i){
            ptr2 = ptr2.next;
        }
        
        if(ptr2 == null){
            return head.next;
        }
        
        while(ptr2.next != null){
            ptr1 = ptr1.next;
            ptr2 = ptr2.next;
        }
        
        ptr1.next = ptr1.next.next;
        return head;
    }
}

File: Remove Nth Node From End of List/Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        if(!head || !n){
            return head;
        }
        ListNode* fakeHead = new ListNode(0);
        fakeHead->next = head;
        ListNode* fast = head;
        ListNode* slow = fakeHead;
        // Steps that fastptr advances slowptr
        for(int i = 0; i < n - 1; ++i){
            fast = fast->next;
            if(!fast){
                return head;
            }
        }
        while(fast->next){
            slow = slow->next;
            fast = fast->next;
        }
        slow->next = slow->next->next;
        return fakeHead->next;
    }
};
